sphinx/builders/singlehtml.py

```diff
@@ -88,13 +88,31 @@ def _get_local_toctree(
         )
         return self.render_partial(toctree)['fragment']
 
+    def ensure_fully_qualified_refids(self, tree: nodes.document) -> None:
+        """Append docname to refids and ids using format
+        document-<docname>#<id>. Compensates for loss of the pathname section
+        of the href, that ensures uniqueness in the html builder.
+        """
+        for node in tree.findall(nodes.Element):
+            doc = node.document
+            if doc is None:
+                continue
+            env = doc.settings.env
+            if 'refid' in node or 'ids' in node:
+                docname = env.path2doc(doc['source'])
```

Comment:
Is there any situation where `node.document` can be empty?  If not, then we can simplify:

```suggestion
            if 'refid' in node or 'ids' in node:
                docname = self.env.path2doc(node.document['source'])
```

Reply:
mypy complains
```
sphinx/builders/singlehtml.py:99:36: error: Value of type "document | None" is not indexable  [index]
```
I tried the NodeMatcher.findall but got the same result. 
With cast 
```
doc = cast(nodes.document, node.document)
```
I get
```
sphinx/builders/singlehtml.py:97:19: error: Redundant cast to "document"  [redundant-cast]
```

```assert node.document is not None``` works

Reply:
Nice, thank you :+1: